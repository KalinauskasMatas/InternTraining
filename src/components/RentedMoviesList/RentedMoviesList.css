.your-movies {
  background-color: var(--bright-background-color);
  margin: 14px 47px 106px 43px;
  padding-bottom: 20px;
  border-radius: 20px;
}

.your-movies-header {
  margin: 0;
  padding: 16px 0 6px 0;
  font-weight: 400;
  letter-spacing: -1px;
  margin-bottom: 23px;
}

.your-movies table {
  padding: 0px 50px 30px 30px;
  border-spacing: 0 10px;
  width: 100%;
}

.your-movies tr:nth-child(n+2) {
  background-color: var(--background-color);
}

.your-movies th {
  padding: 0 23px;
  text-align: left;
}

.your-movies table tr .movie-name {
  border-top-left-radius: 18px;
  border-bottom-left-radius: 18px;
  padding-left: 17px;
}

.your-movies .time-settings {
  background-color: var(--bright-background-color);
}

.time-settings .time-control {
  background-color: var(--dark-button-color);
  color: var(--bright-text-color);
  padding: 0 4px;
  cursor: pointer;
}

.time-settings .time-input {
  padding: 0 10px;
  font-size: 14px;
  font-weight: 700;
}

.your-movies .remove-button {
  background-color: var(--red-button-color);
  color: var(--bright-text-color);
  padding: 10px 8px;
  cursor: pointer;
}

.your-movies td:not(.remove-button), .your-movies th {
  padding-left: 16px;
  text-align: left;
}