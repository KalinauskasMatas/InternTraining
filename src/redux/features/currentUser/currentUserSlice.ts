import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { RootState } from "../../store";
import { UserState } from "../../interfaces";

const initialState: UserState =
  JSON.parse(localStorage.getItem("currUser") as string) || {};

const currentUserSlice = createSlice({
  name: "currentUser",
  initialState,
  reducers: {
    setUser: (state, action: PayloadAction<UserState>) => {
      state.fname = action.payload.fname;
      state.surname = action.payload.surname;
      state.email = action.payload.email;
      state.password = action.payload.password;
      localStorage.setItem("currUser", JSON.stringify(state));
    },
    clearUser: (state) => {
      localStorage.removeItem("currUser");
      state = { ...initialState };
      console.log(state);
    },
  },
});

export const { setUser, clearUser } = currentUserSlice.actions;

export const selectCurrUser = (state: RootState) => state.currentUser;

export default currentUserSlice.reducer;
