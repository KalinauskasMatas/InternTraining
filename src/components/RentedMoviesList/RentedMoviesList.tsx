import { useAppDispatch, useAppSelector } from "../../redux/hooks";

import { updateCurrUser } from "../../redux/features/currentUser/currentUserSlice";
import { useGetAvailableMoviesQuery } from "../../redux/features/api/apiSlice";

import "./RentedMoviesList.css";
import axiosFetch from "../../utils/axiosFetch";

const RentedMoviesList = () => {
  const currentUser = useAppSelector((state) => state.currentUser);
  const dispatch = useAppDispatch();

  const { data: availableMovies } = useGetAvailableMoviesQuery();

  const fullInfo = currentUser.rentMovies.map((movie) => {
    const foundMovie = availableMovies?.find((mov) => movie.id === mov._id);
    return {
      id: movie.id,
      name: foundMovie?.name,
      time: movie.time,
      genre: foundMovie?.genre,
      price: foundMovie?.rentalPrice,
    };
  });

  const handleChangeTime = async (
    id: string,
    time: number,
    changeTime: string
  ) => {
    if (
      (changeTime === "decrease" && time === 12) ||
      (changeTime === "increase" && time === 168)
    )
      return;

    const { data: updatedUser } = await axiosFetch("/movies/time", "PUT", {
      movieId: id,
      changeTime,
    });

    dispatch(updateCurrUser(updatedUser));
  };

  const handleRemove = async (id: string) => {
    const { data: updatedUser } = await axiosFetch("/movies/return", "POST", {
      movieId: id,
    });

    dispatch(updateCurrUser(updatedUser));
  };

  return (
    <div className="body-wrapper">
      <section className="your-movies">
        <h2 className="your-movies-header">Your movies</h2>
        <table id="your-movies-table">
          <tbody>
            <tr>
              <th>Name</th>
              <th>Genre</th>
              <th>Time</th>
              <th>Price</th>
              <th></th>
            </tr>
            {fullInfo.map((movie) => (
              <tr key={movie.id}>
                <td className="movie-name">{movie.name}</td>
                <td>{movie.genre}</td>
                <td>
                  <span className="time-settings">
                    <span
                      className="time-control"
                      onClick={() =>
                        handleChangeTime(movie.id, movie.time, "decrease")
                      }
                    >
                      &#60;
                    </span>
                    <span className="time-input">{movie.time}h</span>
                    <span
                      className="time-control"
                      onClick={() =>
                        handleChangeTime(movie.id, movie.time, "increase")
                      }
                    >
                      &#62;
                    </span>
                  </span>
                </td>
                <td>{movie.price}$</td>
                <td
                  className="remove-button"
                  onClick={() => handleRemove(movie.id)}
                >
                  Remove
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </section>
    </div>
  );
};

export default RentedMoviesList;
