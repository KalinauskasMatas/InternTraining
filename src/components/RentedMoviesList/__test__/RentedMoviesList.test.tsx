import RentedMoviesList from "../RentedMoviesList";

import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";

import renderer from "react-test-renderer";

import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import currentUserSlice from "../../../redux/features/currentUser/currentUserSlice";

const store = configureStore({
  reducer: {
    currentUser: currentUserSlice,
  },
});

afterEach(cleanup);

it("renders without crashing", () => {
  render(
    <Provider store={store}>
      <RentedMoviesList />
    </Provider>
  );
});

it("matches snapshot", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <RentedMoviesList />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
