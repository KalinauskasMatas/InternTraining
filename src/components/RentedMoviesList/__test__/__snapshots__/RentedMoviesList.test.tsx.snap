// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="body-wrapper"
>
  <section
    className="your-movies"
  >
    <h2
      className="your-movies-header"
    >
      Your movies
    </h2>
    <table
      id="your-movies-table"
    >
      <tbody>
        <tr>
          <th>
            Name
          </th>
          <th>
            Genre
          </th>
          <th>
            Time
          </th>
          <th>
            Price
          </th>
          <th />
        </tr>
      </tbody>
    </table>
  </section>
</div>
`;
